lib/plugin-loader.js

```diff
@@ -752,71 +756,275 @@ PluginLoader.prototype = {
+                //DependencyGraph.statuses));
+        newPlugins.push(Object.assign(rejectedPlugin, {error: rejectionError}));
+      }
+  
+      let isFirstRun = Object.keys(this.pluginMap).length === 0;
+      
+      for (const pluginDef of sortedAndRejectedPlugins.plugins) { 
+        try {
+          if (this.pluginMap[pluginDef.identifier]) {
+            bootstrapLogger.warn(`ZWED0034W`, plugin.identifier); //bootstrapLogger.warn(`Skipping install of plugin due to existing plugin with same id=${plugin.identifier}`);
+  
+             continue;
+          }
+          const pluginConfiguration = configService.getPluginConfiguration(
+            pluginDef.identifier, pluginDef.location,
+            this.options.serverConfig, this.options.productCode);
+            bootstrapLogger.debug("ZWED0165I", pluginDef.identifier, JSON.stringify(pluginConfiguration)); //bootstrapLogger.debug(`For plugin with id=${pluginDef.identifier}, internal config` 
+                                  //+ ` found=\n${JSON.stringify(pluginConfiguration)}`);
+  
+          const plugin = makePlugin(pluginDef, pluginConfiguration, pluginContext,
+                                    false, this.options.langManagers);
+          if (plugin) {
+            bootstrapLogger.info("ZWED0124I", plugin.identifier, plugin.location); //bootstrapLogger.log(bootstrapLogger.INFO,
+              //`Plugin ${plugin.identifier} at path=${plugin.location} loaded.\n`);
+              bootstrapLogger.debug("ZWED0164I", plugin.toString()); //bootstrapLogger.debug(' Content:\n' + plugin.toString());
+            let frozen = zluxUtil.deepFreeze(plugin);
+            this.plugins.push(frozen);
+            newPlugins.push(frozen);
+            this.pluginMap[plugin.identifier] = plugin;
+            successCount++;
+          } else {
+            bootstrapLogger.info("ZWED0125I", pluginDef.identifier); //bootstrapLogger.log(bootstrapLogger.INFO,
+              //`Plugin ${pluginDef.identifier} not loaded`);
+            this.pluginMap[pluginDef.identifier] = {}; //mark it as having failed so it will not be retried
+          }
+        } catch (e) {
+          bootstrapLogger.warn("ZWED0035W", pluginDef.identifier, e);
+          //downstream will pick up error
+          newPlugins.push(Object.assign(pluginDef, {error:e}));
+        }
+      }
+      this.registerStaticPluginsWithManagers(sortedAndRejectedPlugins.plugins);
+      for (const plugin of newPlugins) {
+        this.emit('pluginFound', {
+          data: plugin,
+          count: newPlugins.length
+        });
+      }
+  },
+
+  // Note - Not to be confused with auth capabilities, that describe what an auth plugin can do
+  async getComponentCapabilities(config) {
+    let appServerComp = {};
+    let agentComp = {};
+    let host;
+    let port;
+
+    appServerComp.os = process.platform; // Operating system
+    appServerComp.cpu = process.arch; // CPU architecture
+
+    if (config && config.agent.host) {
```

Comment:
I think 843-848 can be refactored as well. If the app server started, you can assume it received a valid host and port for ZSS. The only instance where one wouldnt be set is if someone was using trivial auth, in which case this endpoint/functionality would be useless.

This just looks messy more so than anything, the above criticism could be a bit overstated

Reply:
"someone was using trivial auth, in which case this endpoint/functionality would be useless" Yes, useless for the agent environment, but not useless for the app server environment that uses Node commands instead of auth. At least in theory, it could be such that app server has a plugin that doesn't use the mainframe but only runs on a certain OS, maybe like a web Zowe game that uses a specific type of web framework (i.e. MacOS can't do Flash etc) I think I'll keep this one for better safe than sorry, agree that it looks messy though